packages/react-theme-provider/src/components/ThemeProvider/useThemeProvider.ts

```diff
@@ -35,20 +37,9 @@ export const useThemeProvider = (
   const parentTheme = useTheme();
 
   const theme = mergeThemes(parentTheme, state.theme ?? {});
-  const style = React.useMemo<React.CSSProperties>(() => {
-    // TODO: should we consider insertion to head?
-    //       - how to modify, remove styles?
-    //       - SSR rendering
-
-    // TODO: what variables should be rendered? Merged or only changed?
-    // TODO: how we will proceed with Portals?
-    return {
-      ...state.style,
-      ...themeToCSSVariables(theme),
-    };
-  }, [state.style, theme]);
-
-  Object.assign(state, { theme, style });
+  const themeClassName = useThemeStyleTag({ theme, targetDocument: state.targetDocument });
```

Comment:
Still the same question as in the removed comment - in case of nested providers, do we want the class of the inner provider to only contain changed/added css variables or all variables?

Reply:
Each ThemeProvider would need to consume the CSS classname of the provider above and concat before passing all of it down. It makes sense.

This would involve flattening both parent and merged theme and running a comparison. I think the perf needs to be measured in this case. (I will try it, since I still need to measure perf of this solution vs previous inline style)

```tsx
<ThemeProvider> // Has all the variables for rule `.themeprovider-1`
  <ThemeProvider> // Has a subset of changed variables for rule `.themeprovider-2` passes down `themeprovider-1 themeprovider-2` 
     <Portal  /> // Needs to use `.themeprovider-1 .themeprovider-2`
  </ThemeProvider>
</ThemeProvider>
```

Reply:
> Needs to use `.themeprovider-1 .themeprovider-2`

It depends on how we structure the portal containers - If I have two sibling Providers, we will create two sibling containers in `body` to host all their portals, correct?

What if I have nested Providers? Are the containers in body siblings or nested?

Reply:
In a nested case, the only sane option IMO would be to share the same Portal context and use the same containers.

In a sibling case, similarly, there would be no way for siblings to know about each others' containers, so each sibling provider creates its own portals.

I can't see a way to 'share' css vars for sibling providers.... would you have an idea to handle that ?

Reply:
As documented in this RFC #17815 the perf increase of the current solution is ~40%. I've measured the further optimizations we can take, but they do add extra complexity and might not be too useful considering how many theme providers should be rendered